android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java

```diff
@@ -18,145 +19,244 @@
+			String lowerCaseQuery = query.toLowerCase().trim();
+
+			// Execute query using rawQuery() in order to receive results.
+			if (lowerCaseQuery.startsWith("select") || lowerCaseQuery.startsWith("insert")
+				|| lowerCaseQuery.startsWith("update") || lowerCaseQuery.startsWith("delete")
+				|| (lowerCaseQuery.startsWith("pragma") && !lowerCaseQuery.contains("="))) {
+
+				// Query parameters must be strings.
+				ArrayList<String> parameters = new ArrayList<>();
+				for (Object parameter : parameterObjects) {
+					parameters.add(TiConvert.toString(parameter));
 				}
-				c = db.rawQuery(sql, selectArgs);
-				if (c != null) {
-					// Most non-SELECT statements won't actually return data, but some such as
-					// PRAGMA do. If there are no results, just return null.
-					// Thanks to brion for working through the logic, based off of commit
-					// https://github.com/brion/titanium_mobile/commit/8d3251fca69e10df6a96a2a9ae513159494d17c3
-					if (c.getColumnCount() > 0) {
-						rs = new TiResultSetProxy(c);
-						if (rs.isValidRow()) {
-							rs.next(); // Position on first row if we have data.
+
+				cursor = db.rawQuery(query, parameters.toArray(new String[0]));
+				if (cursor != null) {
+
+					// Validate and set query result.
+					if (cursor.getColumnCount() > 0) {
+						result = new TiResultSetProxy(cursor);
+						if (result.isValidRow()) {
+							result.next();
 						}
-					} else {
-						c.close();
-						c = null;
-						rs = null;
 					}
-				} else {
-					// Leaving for historical purposes, but walking through several different
-					// types of statements never hit this branch. (create, drop, select, pragma)
-					rs = new TiResultSetProxy(null); // because iPhone does it this way.
 				}
+
+				// Query does not return result, use execSQL()
 			} else {
-				Object[] newArgs = null;
-				if (sqlArgs != null) {
-					newArgs = new Object[sqlArgs.length];
-					for (int i = 0; i < sqlArgs.length; i++) {
-						if (sqlArgs[i] instanceof TiBlob) {
-							newArgs[i] = ((TiBlob) sqlArgs[i]).getBytes();
-						} else {
-							newArgs[i] = TiConvert.toString(sqlArgs[i]);
-						}
-					}
-				}
-				db.execSQL(sql, newArgs);
+				db.execSQL(query, parameterObjects);
 			}
-		} catch (SQLException e) {
-			String msg = "Error executing sql: " + e.getMessage();
-			Log.e(TAG, msg, e);
-			if (c != null) {
+		} finally {
+
+			// Cleanup result.
+			if (cursor != null) {
 				try {
-					c.close();
-				} catch (SQLException e2) {
-					// Ignore
+					cursor.close();
+				} catch (Exception e) {
+					// Ignore...
 				}
 			}
-			throw e;
 		}
 
-		return rs;
+		return result;
+	}
+
+	/**
+	 * Asynchronously execute a single SQL query.
+	 * @param callback Result callback for query execution.
+	 * @param query SQL query to execute on database.
+	 * @param parameters Parameters for `query`
+	 */
+	@Kroll.method
+	public void executeAsync(final KrollFunction callback, final String query, final Object... parameters)
```

Comment:
So in JS do we actually have to pass in the callback as the first argument here, like so?
```javascript
const db = Ti.Database.open('whatever');
db.executeAsync(result => { Ti.API.info(result); }, 'SELECT * FROM TABLE WHERE name = ?', myName);
```

Because if so, that is very odd and we should avoid it if possible. I'm assuming the use of the varargs for parameters here is why the callback needed to come first? Can we re-arrange it so the callback is the last JS arg? If we can't "natively" with our binding layer, we could add JS code to basically do the swap under the hood in the same way we do some hacking of say Ti.App.Properties here: https://github.com/appcelerator/titanium_mobile/blob/master/android/modules/app/src/js/properties.js

Reply:
Second, it's also not idiomatic to have the async variant have the extra name suffix.

Assuming we can hack around the arguments in JS or whatever, why not just use the same method name and if the last argument is a callback function then treat it as async, otherwise sync?

I suppose that boxes us in if we want to return a Promise...

Maybe introduce a db.executeSync, db.executeAsync where they are explicitly sync/async, have the existing method do either sync or async based on if callback is passed in, and spit out a deprecation warning if no callback is passed to it (basically saying, hey in the future this will return a Promise - so if you want sync behavior use db.executeSync)? I think at some point we want to remove the default sync nature of the method and push people towards async naturally - and eventually support Promises.

Reply:
I've modified the parameters so the callback is the last parameter, we need to keep `Ti.Database.execute()` as is to not cause breaking changes.

```
Ti.Database.DB.executeAsync(query, parameters[], result => { ... });

results[] = Ti.Database.DB.executeAll(queries[]);
Ti.Database.DB.executeAllAsync(queries[], results[] => { ... });
```

I'd prefer to introduce as few API changes possible until we make the jump to Promise based APIs in the near future. I didn't intend to refactor all of our Ti.Database.DB APIs with this PR.

I think these APIs allow the flexibility to do the primary things, including transactions if the user chooses to do so. We should create a ticket to refactor the methods in `9.0.0`?